# Exploring Hackers News Posts
In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

### Introduction
First, we'll read in the data and remove the headers.

In [1]:
from csv import reader
hn=list(reader(open("hacker_news.csv")))
print(hn[:6])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['11966167', 'UK votes to leave EU', 'http://www.bbc.co.uk/news/uk-politics-36615028', '3125', '2531', 'dmmalam', '6/24/2016 3:48'], ['12445994', 'iPhone 7', 'http://www.apple.com/iPhone7', '756', '1733', 'benigeri', '9/7/2016 18:52'], ['11807450', 'Moving Forward on Basic Income', 'http://blog.ycombinator.com/moving-forward-on-basic-income', '1330', '1448', 'dwaxe', '5/31/2016 16:20'], ['10982340', 'Request For Research: Basic Income', 'https://blog.ycombinator.com/basic-income', '1876', '1120', 'mattkrisiloff', '1/27/2016 19:23'], ['11814828', 'Ask HN: Who is hiring? (June 2016)', '', '644', '1007', 'whoishiring', '6/1/2016 15:01']]


In [2]:
header=hn[0]
header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [3]:
hn=hn[1:]
hn[:6]

[['11966167',
  'UK votes to leave EU',
  'http://www.bbc.co.uk/news/uk-politics-36615028',
  '3125',
  '2531',
  'dmmalam',
  '6/24/2016 3:48'],
 ['12445994',
  'iPhone 7',
  'http://www.apple.com/iPhone7',
  '756',
  '1733',
  'benigeri',
  '9/7/2016 18:52'],
 ['11807450',
  'Moving Forward on Basic Income',
  'http://blog.ycombinator.com/moving-forward-on-basic-income',
  '1330',
  '1448',
  'dwaxe',
  '5/31/2016 16:20'],
 ['10982340',
  'Request For Research: Basic Income',
  'https://blog.ycombinator.com/basic-income',
  '1876',
  '1120',
  'mattkrisiloff',
  '1/27/2016 19:23'],
 ['11814828',
  'Ask HN: Who is hiring? (June 2016)',
  '',
  '644',
  '1007',
  'whoishiring',
  '6/1/2016 15:01'],
 ['11049067',
  'GitHub is undergoing a full-blown overhaul as execs and employees depart',
  'http://www.businessinsider.com/github-the-full-inside-story-2016-2',
  '808',
  '973',
  'easyd',
  '2/6/2016 18:43']]

In [4]:
ask_posts,show_posts,other_posts=[],[],[]
for items in hn:
    if items[1].lower().startswith("ask hn"):
        ask_posts.append(items)
    elif items[1].lower().startswith("show hn"):
        show_posts.append(items)
    else:
        other_posts.append(items)

In [5]:
total_ask_comments=0
for items in ask_posts:
    total_ask_comments+=int(items[4])

avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

74.0994652406417


In [6]:
total_show_comments=0
for items in show_posts:
    total_show_comments+=int(items[4])
print(total_show_comments/len(show_posts))

49.34550561797753


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

### Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [7]:
import datetime as dt

result_list=[]
for items in ask_posts:
    result_list.append([items[6],int(items[4])])
    
counts_by_hour={}
comments_by_hour={}
for item in result_list:
    hour=dt.datetime.strptime(item[0],"%m/%d/%Y %H:%M")
    hour=dt.datetime.strftime(hour,"%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=int(item[1])
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=int(item[1])


Sorting and Printing Values from a List of Lists

In [8]:
avg_by_hour=[]
for item,num_comment in counts_by_hour.items():
    avg_by_hour.append([item,round(comments_by_hour[item]/num_comment,2)])
    
swap_avg_by_hour=[]
for hr,num_comment in avg_by_hour:
    swap_avg_by_hour.append([num_comment,hr])
    
sorted_swap=sorted(swap_avg_by_hour,reverse=True)

In [11]:
for cm,hr in sorted_swap[:5]:
    print("{h} {c} average comments per post.".format(h=dt.datetime.strftime(dt.datetime.strptime(hr,"%H"),"%H:%M"),c=cm))

15:00 190.91 average comments per post.
13:00 95.14 average comments per post.
02:00 76.69 average comments per post.
05:00 73.18 average comments per post.
12:00 70.73 average comments per post.


The hour that receives the most comments per post on average is 15:00, with an average of 190.91 comments per post. There's about a 100% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.